feat(backup/s3): optionally compress backup contents

oleschoenburg opened the pull request:
Introduces a new configuration option for the S3 backup store: `ZEEBE_BROKER_DATA_BACKUP_S3_ENABLECOMPRESSION`. For now, this is disabled by default.

When enabled, files above a hard coded threshold (currently 8 MiB) are compressed with [Zstandard](https://github.com/facebook/zstd). We first compress to a temporary file before uploading. In-memory or streaming compression is impractical. For each file, the used compression algorithm is stored as metadata in the manifest.

On restore, files that have a compression algorithm stored in the metadata are downloaded to a temporary file and then decompressed to the actual target path.

Compression is handled by [commons-compress](https://commons.apache.org/proper/commons-compress/index.html) which provides a consistent interface for many compression algorithms. If we decide to change the default compression algorithm or make it configurable, restoring will use whatever algorithm was used to compress, thus achieving backwards compatibility. 

Currently, the compression algorithm is hard coded as Zstandard. This requires an additional dependency on the native wrapper library which provides the binaries for all supported architectures: https://github.com/luben/zstd-jni#binary-releases

## Alternatives
At first I considered archiving and compressing the entire backup or at least all segment and snapshot files together. However, this could easily exceed the current limitation of 5 GiB per file upload. It's not too bad though, because this way we can actually compress files in parallel and don't increase the peak disk usage by too much.

Another alternative was to implement this independently from the backup store and let the Zeebe broker handle compression. I believe handling it in the store itself makes more sense as some backends may have support for native compression. Additionally, letting the broker handle compression would again increase peak disk usage. Worst case, if backup content is not at all compressible, this could double the required disk size.

## Improvements

We should consider using a dedicated thread pool or another mechanism to control how many files are (de-)compressed in parallel. We don't have any data yet, but it is easy to imagine that doing a lot of compression in parallel could have a considerable impact on CPU usage, disk usage and disk I/O. 
Actually, it might make sense to use such a mechanism more broadly on parallel upload and download of files to also control the impact on network I/O. 

Some parts of the backup might not be very compressible. In particular, parts of the snapshot are already compressed by RocksDB. It might make sense to not even attempt to compress snapshot files.

closes #10846

deepthidevaki commented on backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/FileSetManager.java:
```diff
@@ -60,16 +132,72 @@ CompletableFuture<NamedFileSet> restore(
+        final var decompressedOutput =
+            new CompressorStreamFactory().createCompressorInputStream(algorithm, input)) {
+      IOUtils.copy(decompressedOutput, output);
+      if (LOG.isTraceEnabled()) {
```
:wrench: Why would you add this check, but not in other places? I don't think this is required.

deepthidevaki replied:
:+1:

deepthidevaki commented on broker/src/main/java/io/camunda/zeebe/broker/system/configuration/backup/S3BackupStoreConfig.java:
```diff
@@ -20,6 +20,7 @@ public class S3BackupStoreConfig implements ConfigurationEntry {
   private String secretKey;
   private Duration apiCallTimeout = Duration.ofSeconds(180);
   private boolean forcePathStyleAccess = false;
+  private boolean enableCompression = false;
```
:question:  Do you think we will have to support other compression algorithms in future? If yes, for future extendibility, instead of a boolean flag, it would be better to specify the compression algorithm here. See `ZEEBE_BROKER_CLUSTER_MESSAGECOMPRESSION` as an example. 
```
ZEEBE_BROKER_DATA_BACKUP_S3_COMPRESSION = none | zstd
```
This way we can add new compression algorithms without changing the configuration format. What do you think?

NOTE: It is not clear to me if there is value in letting users choose the compression algo. So I'm ok to leave it as it is.

deepthidevaki commented on dist/src/main/config/broker.standalone.yaml.template:
```diff
@@ -278,6 +278,10 @@
           # This setting can also be overridden using the environment variable ZEEBE_BROKER_DATA_BACKUP_S3_FORCEPATHSTYLEACCESS
           # forcePathStyleAccess: false
 
+          # When enabled, parts of the backup may be compressed.
```
```suggestion
          # When enabled, parts of the backup may be compressed. This helps to reduce the required storage space in s3 store. Enabling compression may use more cpu and disk while taking the backup.
```
Something along these line to indicate the pros/cons of enabling this would be useful.

deepthidevaki commented on qa/integration-tests/src/test/java/io/camunda/zeebe/it/backup/BackupMultiPartitionTest.java:
```diff
@@ -114,7 +114,8 @@ void createBackupStoreForTest() {
             Duration.ofSeconds(15),
-            true);
+            true,
+            false);
```
:upside_down_face: Number of params to the constructur is increasing. A builder for the config might be better.

oleschoenburg commented:
@deepthidevaki Thanks for the review :+1:
I've rebased to resolve merge conflicts and added a [commit](https://github.com/camunda/zeebe/pull/11017/commits/e06f95e18a1f4bbe4eec406d219b5c993cee0366) that allows users to configure the compression algorithm to use.

oleschoenburg commented:
bors r+
